src/core/SampleBuffer.cpp

```diff
@@ -1447,3 +1457,167 @@ SampleBuffer::handleState::~handleState()
+
+
+
+// Fading related functions
+
+float SampleBuffer::fadingVal(f_cnt_t pos, bool is_left)
+{
+    float relativePos = static_cast<float>(pos)/static_cast<float>(m_frames);
+    return (this->*fadefuncPtr)(relativePos, is_left);
+}
+
+
+void SampleBuffer::setLeftFader(float lfader)
+{
+    m_leftFader = lfader;
+    checkFadingActive();
+}
+
+
+void SampleBuffer::setRightFader(float rfader)
+{
+    m_rightFader = rfader;
+    checkFadingActive();
+}
+
+
+void SampleBuffer::checkFadingActive()
+{
+    if ((m_leftFader > 0.0) || (m_rightFader < 1.0))
+    {
+        m_fading = true;
+    }
+    else
+    {
+        m_fading = false;
+    }
+}
+
+
+float SampleBuffer::bezierFade(float relativePos, bool is_left)
+{
+    float amp = 1.0f;
+    if (is_left)
+    {
+        amp *= m_leftBezierFade.evaluate(relativePos/m_leftFader);
+    }
+    else
+    {
+        amp *= m_rightBezierFade.evaluate(1-(relativePos-m_rightFader)/(1-m_rightFader));
+    }
+    return amp;
+}
+
+
+float CubicBezier::cubicBezierLastSegmentP1(float t, float p1, float p2)
+{
+    return 2*p1*t*(1-t) + p2*pow(t, 2);
+}
+
+
+float CubicBezier::cubicBezierLastSegmentP2(float t, float p1, float p2)
+{
+    return p1*pow(1-t, 2) + 2*p2*t*(1-t) + pow(t, 2);
+}
+
+
+float CubicBezier::cubicBezier(float t, float p1, float p2)
+{
+    return (cubicBezierLastSegmentP1(t, p1, p2) * (1-t)
+            + cubicBezierLastSegmentP2(t, p1, p2) * t);
+}
+
+
+float CubicBezier::cubicBezierDerivative(float t, float p1, float p2)
+{
+    return 3*(3*pow(t, 2)-4*t+1)*p1 + 3*t*(2-3*t)*p2 + 3*pow(t, 2);
+}
+
+
+float CubicBezier::solveXforTNewton(float x, bool useCache=true)
+{
+    int max_iter = 100;
+    float t, t_up, f, f_prime;
+    if (useCache)
+    {
+        t = m_tCache;
+    }
+    else
+    {
+        t = 0.5f;
+    }
+    for (int i=0; i<max_iter; ++i)
+    {
+        f = cubicBezier(t, m_P1.x(), m_P2.x()) - x;
+        f_prime = cubicBezierDerivative(t, m_P1.x(), m_P2.x());
+        t_up = t - f/f_prime;
+
+        if (abs(t_up - t) < 10e-3)
+        {
+            break;
+        }
+        t = t_up;
+    }
+    if (useCache)
+    {
+        m_tCache = t_up;
+    }
+    return t_up;
+}
+
+
+float CubicBezier::solveXforTBisection(float x)
+{
+    float a = 0.0f;
+    float b = 1.0f;
+    float t, f_a, f_t;
+
+    int max_iter = 10;
+    for (int i = 0; i < max_iter; ++i)
+    {
+        t = (a + b)/2;
+        f_t = cubicBezier(t, m_P1.x(), m_P2.x()) - x;
+        f_a = cubicBezier(a, m_P1.x(), m_P2.x()) - x;
+        bool a_sign = (f_a >= 0) ? true: false;
+        bool t_sign = (f_t >= 0) ? true: false;
+        if (a_sign == t_sign)
+        {
+            a = t;
+
+        }
+        else
+        {
+            b = t;
+        }
+    }
+    return t;
+}
+
+
+float CubicBezier::evaluate(float x)
+{
+    float t = solveXforT(x);
+    return cubicBezier(t, m_P1.y(), m_P2.y());
+}
+
+
+float SampleBuffer::fadefunc(f_cnt_t pos)
+{
+    // This function should be cached in the future
```

Comment:
```suggestion
    // TODO: This function should be cached in the future
```

Makes it easier to find places where enhancements are needed.